Document that reduce-like functions use `init` one or more times

mbauman opened the pull request:
OK, here's take two, building on the work that was done in #53871.  We preserve the neutral element requirement but guarantee that `init` is used to start all evaluation chains.

The news summarizes as succinctly as I can:

> `init` is now guaranteed to be used for non-empty collections one or more times. This ensures that all calls
  to the reducing function `op` have one argument that is either `init` or the result from a
  previous `op` evaluation. Previously `init` was explicitly allowed to be omitted from the reduction entirely.

And the docs for `mapreduce` say a bit more:

> If provided, `init` is the return value for empty collections and is used one or more times as
an argument to `op` for non-empty collections. The `init` value is not transformed by the function `f`.
Using `init` ensures that all calls to `op` have one argument that is either `init`
or the result from a previous `op` evaluation, and the ordering of these arguments is
unspecified. As it may appear in the reduction one or more times, it must be a neutral element for `op` that
does not change the result by being used more than once. It is generally an error to call `mapreduce`
with empty collections without specifying an `init` value, but in unambiguous cases an
identity value for `op` may be returned; see [`Base.reduce_empty`](@ref) for more details.

Closes https://github.com/JuliaLang/julia/issues/49042.

mbauman committed:
Add more clarity to `init`'s role in reduce and mapreduce

mbauman committed:
fix too many dots......

Co-authored-by: mikmoore <95002244+mikmoore@users.noreply.github.com>

mbauman committed:
Incorporate feedback; define left-most position for init

mbauman committed:
Apply suggestions from code review

Co-authored-by: Steven G. Johnson <stevenj@alum.mit.edu>

mbauman committed:
Add NEWS.md

mbauman committed:
typo fix

mbauman committed:
Revert back to the neutral element init

and add more clarifications

mbauman committed:
update NEWS

mbauman commented on base/reduce.jl:
```diff
@@ -280,29 +280,65 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+Apply function `f` to each element(s) in `itrs`, and then repeatedly call the 2 argument
+function `op` with those results or results from previous `op` evaluations until a single value is returned.
+
+If provided, `init` provides the return value for empty collections and is used one or more times as
```
```suggestion
If provided, `init` is the return value for empty collections and is used one or more times as
```

mbauman committed:
minor wording

mbauman commented on base/reduce.jl:
```diff
@@ -280,29 +280,65 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+an argument to `op` for non-empty collections. The `init` value is not transformed by the function `f`.
+Using `init` ensures that all calls to `op` take one argument from the (mapped) collection(s) and the other
+from either `init` or the result from a previous `op` evaluation, and the ordering of these arguments is
+unspecified. As it may appear in a reduction one or more times, it must be a neutral element for `op` that 
```
```suggestion
unspecified. As it may appear in the reduction one or more times, it must be a neutral element for `op` that 
```

mbauman committed:
more minor wording

mbauman committed:
Propagate changes to other reduce family members

jariji commented:
Do we want to specify whether it's `op(init, _)` or `op(_, init)`?

mbauman commented on base/reduce.jl:
```diff
@@ -280,29 +280,65 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+an argument to `op` for non-empty collections. The `init` value is not transformed by the function `f`.
+Using `init` ensures that all calls to `op` take one argument from the (mapped) collection(s) and the other
+from either `init` or the result from a previous `op` evaluation, and the ordering of these arguments is
+unspecified. As it may appear in the reduction one or more times, it must be a neutral element for `op` that 
```
```suggestion
unspecified. As it may appear in the reduction one or more times, it must be a neutral element for `op` that
```

mbauman commented on base/reduce.jl:
```diff
@@ -452,36 +488,52 @@ _mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted) = mapfoldl(f,
+an argument to `op` for non-empty collections.
+Using `init` ensures that all calls to `op` take one argument from `itr` and the other
+from either `init` or the result from a previous `op` evaluation, and the ordering of these arguments is
+unspecified. As it may appear in a reduction one or more times, it must be a neutral element for `op` that 
```
```suggestion
unspecified. As it may appear in a reduction one or more times, it must be a neutral element for `op` that
```

mbauman committed:
whitespace

jariji commented on base/reduce.jl:
```diff
@@ -502,7 +554,7 @@ The return type is `Int` for signed integers of less than system word size, and
 
 The value returned for empty `itr` can be specified by `init`. It must be
-the additive identity (i.e. zero) as it is unspecified whether `init` is used
+the additive identity (i.e. zero) as it may be used one or more times
```
"may be" vs "will be"?

mbauman replied:
Good point!

mbauman commented on NEWS.md:
```diff
@@ -77,6 +77,12 @@ New library features
+* The `init` keyword argument in `reduce(op, itr; init)` and other reduction functions with implementation-defined
+  associativity (`mapreduce`, `maximum`, `minimum`, `sum`, `prod`, `any`, and `all`)
+  is now guaranteed to be used for non-empty collections one or more times. This ensures that all calls
+  to the reducing function `op` take one argument from `itr` and the other from either `init` or the result from a
+  previous `op` evaluation. Previously it was explicitly allowed to be omitted from the reduction entirely.
```
```suggestion
  to the reducing function `op` have one argument that is either `init` or the result from a
  previous `op` evaluation. Previously `init` was explicitly allowed to be omitted from the reduction entirely.
```

mbauman committed:
fixup wordings around the arguments to `op`

stevengj commented on base/reduce.jl:
```diff
@@ -280,29 +280,65 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+an argument to `op` for non-empty collections. The `init` value is not transformed by the function `f`.
+Using `init` ensures that all calls to `op` have one argument that is either `init`
+or the result from a previous `op` evaluation, and the ordering of these arguments is
+unspecified. As it may appear in the reduction one or more times, it must be a neutral element for `op` that
```
```suggestion
unspecified. As it may appear in the reduction one or more times, it must be a neutral (identity) element for `op` that
```
I think [identity element](https://en.wikipedia.org/wiki/Identity_element) is the more common name in math?

jonas-schulze replied:
```suggestion
unspecified. As it may appear in the reduction one or more times, it must be a neutral element (identity element) for `op` that
```

to improve `grep`ability?

mbauman replied:
I would rather just go all-in on identity — it is indeed a far more common term.  That's what I've done in my latest push.

jonas-schulze commented:
> Do we want to specify whether it's `op(init, _)` or `op(_, init)`?

I think this is a fair point, especially since `fold[lr]` immediately refer to `reduce`. Either option is fine by me. Using the same order for all three (six including `map*`) makes switching seamless.

It most cases, both arguments will likely have the same type, i.e. `op(::T, ::T) where T`. But surely, there will be some cases where this is not the case, such that it is necessary to specify whether it is `op(acc, val)` or the other way around.

Seelengrab commented:
This docstring feels awfully long, especially when viewed in `help>` or as a function hint in an editor - maybe it's better to have this behavior recorded in the manual, and refer to that section in the docstring?

MasonProtter commented on base/reduce.jl:
```diff
@@ -280,29 +280,65 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
+
+In contrast with [`mapfoldl`](@ref) and [`mapfoldr`](@ref), the sequence of
+function evaluations and the associativity of the reduction is not specified
+and may vary between different methods and across Julia versions. Some implementations
+may reuse the return value of `f` for elements that appear multiple times in the
+collection(s).
```
I'm uneasy with the idea that `f`'s return values may be re-used rather than repeatedly calling `f`. 

It's very normal to have an impure `f` function even for a parallel `mapreduce` so long as calls to `f` don't depend on the order of operations. For example, it's totally normal for people to do  `I/O` in `f`, or they might want to return a mutable collection, where it'd be incorrect and buggy to re-use the same collection.

For example, a parallelizable `map!(f, dest, coll)` can be pretty efficiently implemented as follows:
```julia
function mymap!(f, dest, coll)
    mapreduce((l, r) -> nothing, eachindex(dest, coll)) do i
        dest[i] = f(coll[i])
    end
end
```

mbauman replied:
This is not a new part of the docstring, though.  It dates way back to #10536.

stevengj replied:
Maybe clarify that this normally only happens for "sparse" data structures that implicitly store some repeated element?

MasonProtter replied:
Hm I see. That's a bit scary.

MasonProtter replied:
After some thought, I like @stevengj's suggestion. I think it's a sensible corner case to make it clear to people that we're not normally in the buisness of memoizing reductions, but if a datastructure has its own memoization scheme then we'll reserve the right to also take advantage of that.

mbauman replied:
I'd ask that this be tackled separately — I fiddled around with talking about memoized/sparse/something, but it's so hard to talk about in the abstract.  I've opened issue #55238 to not lose the thread.

MasonProtter reviewed (changes requested):
Looks great, thank you for taking this on @mbauman. I just have one remaining point of disagreement about assuming purity of `f` and then I'm fully on board.

mbauman committed:
use extended help for (map)?reduce, propagate language to sum et al

StefanKarpinski commented:
I'm a bit late to the party, but after consideration and discussion with @mbauman, I'm on board with this change. I think the guarantee can be expressed more clearly this way:

> When no `init` value is provided, some `op` calls will take collection values as both arguments. When `init` is provided, values from the collection will only be used as right arguments to `op`, never as left arguments—the left argument of `op` will always be `init` or the result of a previous call to `op`.

I agree that the wording of the new docs is a bit verbose, but we can always worry about that after the functionality is nailed down.

StefanKarpinski commented:
The argument ordering is very hard to express without explicitly invoking an `op` call tree. Perhaps that's the best way to write it anyway. Something like

> The collection values are guaranteed to appear as leaves in the tree of `op` calls in order, but that does not guarantee that those op calls are evaluated in that order. For example, the left and right halves of an array may be reduced concurrently. When the results of those two reductions are combined the left and right reduced values will respectively be the left and right arguments to the final `op` call.

mbauman committed:
Merge remote-tracking branch 'origin/master' into mb/the-bourne-ultinitum

mbauman commented:
I don't like requiring that `init` needs to be a left-argument to `op` because that means that `foldr` would not be a valid implementation for `reduce`.  But otherwise I agree.  I'm seeing if I can get the description a little more clear with fresh eyes here.

jariji commented:
Is "`init` is the right argument of `op` in `foldr`" documented?

mbauman commented:
As an example, yes.

mbauman committed:
Reword [map]reduce again

Try to make each paragraph here "punchier" and more pointed:
* First paragraph more specifically describes the general behavior
* Second paragraph is about `init`, clarify what it means for `op` args
* Third paragraph is about `op`, but point to extended help
* Fourth paragraph is about empty collections

The extended help now reads clearer with a bulleted list and explicitly mentions argument ordering... and is also a bit more careful about floating point associativity (or the lack thereof) in mapreduce

mbauman commented:
OK, I've pushed a new re-wording. I find this more clear, personally.

We seem have a very solid consensus that "reduce-like functions use init one or more times" so I've removed the RFC from the title.  I think I've faithfully represented that idea in the docs with this latest push.

Of course comments on the wordings are still very welcome, but I am much happier with the state of this now.

mbauman committed:
restore the note on re-use of f's results

mbauman committed:
remove ref to non-public Base.reduce_empty

and try to fix ≈ docs

mbauman committed:
fix typo in reduce @ref

mikmoore commented:
It makes me a tad anxious that we don't guarantee whether the `init` is a first or second argument to `op`. Could there be a situation where this is an issue, where we might need distinct left and right neutral elements? Does a neutral element always commute for an associative operation? This might be a trivial result from group theory, but I don't know the answer.

nsajko commented:
> Does a neutral element always commute for an associative operation?

No: https://en.wikipedia.org/wiki/Semigroup#Identity_and_zero

nsajko commented:
> A two-sided identity (or just identity) is an element that is both a left and right identity. Semigroups with a two-sided identity are called [monoids](https://en.wikipedia.org/wiki/Monoid). A semigroup may have at most one two-sided identity. If a semigroup has a two-sided identity, then the two-sided identity is the only one-sided identity in the semigroup. If a semigroup has both a left identity and a right identity, then it has a two-sided identity (which is therefore the unique one-sided identity).

So a monoid is required, not just associativity.

mbauman commented:
I have two reasons why that's not a major concern in my view.  One is that there are three "types" of arguments that get passed to `op`:
* `init` itself
* values from `itr` — which necessarily could be either left- or right-args (but not both when `init` is provided)
* return values from previous `op` calls — which are also necessarily left- or right-args (or both)

Because the `init` value is also the return value for empty collections, it should be of a similar "type" (yes, I mean that both as a literal `Type` and as a softer "interpretation" or "category of thing") as `op`'s return values.  And `op` necessarily needs to be able to accommodate previous return values in either slot.

The second is more simplistic, but it really seems like `foldr` should be a valid implementation for `reduce`, and `foldr(=>, [1,2,3], init=0)` is (and should be) `1 => (2 => (3 => 0))`.

mbauman commented:
Perhaps more succinctly, just look at `foldl`. With `init`, it will pass *all* its `itr` values as right-args.  All the left args (except one) are results from previous operations.  And `init` is that final left-arg.  Flip the associativity with `-r` and the argument orderings flip.  So `init`'s placement flips.  It makes sense that an arbitrary associativity would grant this leeway.

If care about argument orderings between `init` and an `itr` value, you probably care about the argument orderings between an `op` result and an `itr` value — and so then you care about the associativity.  And you should use `foldl` or `foldr`.

stevengj commented:
When the operation is `+`, should we check `iszero(init)` for backward compatibility and print a deprecation warning if it is `false`?

mbauman commented:
Well the previous documentation explicitly says that it may or may not be used... but it'd probably be a good idea to get an idea of how many folks were relying on the old implementation.  I say we first try the "full pairwise" in #52397 so we can run PkgEval there to gather the data.  A deprecation path would be kind but would probably have to be fairly finely scoped (the generic `iszero` can throw method errors).

mikmoore commented:
One behavior I'd like to preserve is
```julia-repl
julia> sum(fill(-0.0, 0))
0.0

julia> sum(fill(-0.0, 3))
-0.0
```
It seems that the way to get this is to make `sum` on `IEEEFloat` use `init=-0.0` (the `IEEEFloat` additive identity) but allow (the option of) a distinct return value of `+0.0` for the `isempty` case. I hate that this makes the implementation more complicated but I don't see a great way around it.

I would be somewhat unhappy if a sum of `-0.0`s yielded `+0.0` (`sum(x)` would differ from `+(x...)` in a not-just-reassociation way), and I expect the change would break things (both existing and yet-unwritten) in at least a few places, but I know most people would be unhappy if an empty sum yielded `-0.0` (and this would almost certainly break things in many places).

(`+0.0` vs `-0.0` might seem pretty inconsequential until you write `inv(sum(x))`)

mbauman commented:
I've become completely radicalized against initial value guesswork _when it's not specified_, and the proposed behavior change here is only about what to do when `init` _is specified_.

Guessing at `init` is simply _so bug-prone_.  I think the only sensible answer here is _always doing_ what the change in #55318 ensures for dimensional reductions. That is:

> * 0 values:
>    * Have init? Return it. Otherwise:
>    * Return `mapreduce_empty` (usually an error, sometimes an unambiguous value like `0.0`)
> * 1 value (`a`): 
>     * Have init? Return `op(init, f(a))`. Otherwise:
>     * Return `mapreduce_first` (typically `f(a)`)
> * 2+ values (`a1`, `a2`):
>     * Have init? Start every chain with `op(op(init, f(a1)), f(a2))` and continue. Otherwise:
>     * Start every chain with `op(f(a1), f(a2))` and continue

stevengj commented:
What is the status of this PR?

mbauman commented:
I started work on making the implementations follow this, but I ran out of time to accomplish it.

https://github.com/JuliaLang/julia/pull/55318

christiangnrd commented:
I came about this PR from https://github.com/JuliaGPU/AcceleratedKernels.jl/issues/28 -> https://github.com/JuliaFolds2/OhMyThreads.jl/issues/144 -> https://github.com/JuliaLang/julia/issues/49042 and I think the `mapreduce` and `reduce` docstrings defined in reducedims.jl should also be updated to eliminate any ambiguity.
